x/dex/module.go

```diff
@@ -266,38 +260,51 @@ func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.Val
 		cachedCtx, msCached := store.GetCachedContext(ctx)
 		// cache keeper in-memory state
 		memStateCopy := am.keeper.MemState.DeepCopy()
-		finalizeBlockMessages := map[string]*dextypeswasm.SudoFinalizeBlockMsg{}
+		finalizeBlockMessages := sync.Map{} // of type map[string]*dextypeswasm.SudoFinalizeBlockMsg{}
 		for contractAddr := range validContractAddresses {
-			finalizeBlockMessages[contractAddr] = dextypeswasm.NewSudoFinalizeBlockMsg()
+			finalizeBlockMessages.Store(contractAddr, dextypeswasm.NewSudoFinalizeBlockMsg())
 		}
 
-		for contractAddr, contractInfo := range validContractAddresses {
+		validContractInfo := []types.ContractInfo{}
+		for _, contractInfo := range validContractAddresses {
+			validContractInfo = append(validContractInfo, contractInfo)
+		}
+		mu := sync.Mutex{}
+		runnable := func(contractInfo types.ContractInfo) {
 			if !contractInfo.NeedOrderMatching {
-				continue
+				return
 			}
-			ctx.Logger().Info(fmt.Sprintf("End block for %s", contractAddr))
+			cachedCtx.Logger().Info(fmt.Sprintf("End block for %s", contractInfo.ContractAddr))
 			if orderResultsMap, err := am.endBlockForContract(cachedCtx, contractInfo); err != nil {
-				ctx.Logger().Error(fmt.Sprintf("Error for EndBlock of %s", contractAddr))
-				failedContractAddresses.Add(contractAddr)
+				cachedCtx.Logger().Error(fmt.Sprintf("Error for EndBlock of %s", contractInfo.ContractAddr))
+				failedContractAddresses.Add(contractInfo.ContractAddr)
 			} else {
 				for account, orderResults := range orderResultsMap {
 					// only add to finalize message for contract addresses
-					if msg, ok := finalizeBlockMessages[account]; ok {
-						msg.AddContractResult(orderResults)
+					if msg, ok := finalizeBlockMessages.Load(account); ok {
+						typedMsg := msg.(*dextypeswasm.SudoFinalizeBlockMsg)
+						mu.Lock()
+						typedMsg.AddContractResult(orderResults)
+						mu.Unlock()
 					}
 				}
 			}
 		}
+		runner := contract.NewParallelRunner(runnable, validContractInfo)
```

Comment:
nit - can we make this `validContractsInfo` so we know its multiple contracts rather than one

Reply:
updated